File: lab5/bachetsgame/main.cpp
#include <cstdio>
using namespace std;

int main(){
    int n, m;
    int a[10];
    bool arr[1000001];

    while(scanf("%d %d", &n, &m) == 2){
        for(int i=0; i<m; i++){
            scanf("%d", &a[i]);
        }

        arr[0] = false;
        for(int i=1; i <= n; i++){
            arr[i] = false;
            for(int j=0; j<m; j++){
                if(i>=a[j] && !arr[i-a[j]]){
                    arr[i] = true;
                    break;
                }
            }
        }
        if(arr[n]){
            printf("Stan wins\n");
        } else{
            printf("Ollie wins\n");
        }
    }
}

File: lab5/bachetsgame/main.py
import sys
from array import *

line = sys.stdin.readline()
arr = array('b', [False]*(1000001))
while line != '':
    nums = list(map(int, line[:-1].split(' ')))
    moves = nums[2:]
    arr[0] = False
    for i in range(1, nums[0]+1):
        arr[i] = False
        for move in moves:
            if i >= move and not arr[i-move]:
                arr[i] = True
                break
    if arr[nums[0]]:
        print('Stan wins')
    else:
        print('Ollie wins')
    line = sys.stdin.readline()